Several GI related optimizations and fixes

reduz opened the pull request:
-SDFGI direct light is done over many frames
-SDFGI Changed settings for rays/frame
-SDFGI Misc optimizations
-SDFGI Bug fix on probe scroll

-GIProbe was not working, got it to work again
-GIProbe dynamic objects were not working, fixed

-Added a half size GI option.

-Added a --profile-gpu command line option to make it easier to profile graphics from command line.

-All shadow rendering is done with raster now (no compute)
-All shadow rendering is done by rendering directly to the shadow atlas
-Improved how buffer clearing is done to optimize the above.
-16 bits shadow map support, and made the default

reduz committed:
Several GI related optimizations and fixes

-SDFGI direct light is done over many frames
-SDFGI Changed settings for rays/frame
-SDFGI Misc optimizations
-SDFGI Bug fix on probe scroll

-GIProbe was not working, got it to work again
-GIProbe dynamic objects were not working, fixed

-Added a half size GI option.

clayjohn commented on servers/rendering/renderer_rd/shaders/resolve.glsl:
```diff
@@ -58,6 +58,117 @@ void main() {
+	group1.y = texelFetch(source_depth, pos, 1).r;
+
+	//4X
+	if (params.sample_count <= 4) {
```
Shouldnt these be >= ?

Right now a sample count of 2 will run all the conditions, while a sample count of 16 only uses the last

JFonS replied:
Agreed.

JFonS commented on servers/rendering/renderer_rd/shaders/resolve.glsl:
```diff
@@ -58,6 +58,117 @@ void main() {
+		}
+		if (freq.w < min_f) {
+			best_index = 3;
+			min_f = freq.w;
```
Probably compiled away, but `min_f` is not used any more here.

reduz committed:
Shadow map rendering optimization

-All shadow rendering is done with raster now (no compute)
-All shadow rendering is done by rendering directly to the shadow atlas
-Improved how buffer clearing is done to optimize the above.
-Ability to set shadows as 16 bits.

reduz commented:
Do not merge yet

reduz commented:
Ok now can merge if CI passes.

mrjustaguy commented:
This PR Causes a Shadow Bug on Nvidia Quadro K2000 Driver 452.06, W10 64 bit

![Vulkan Shadow Bug](https://user-images.githubusercontent.com/50302479/105700576-56c49400-5f09-11eb-8586-18966d48ec60.PNG)

[Shadow.zip](https://github.com/godotengine/godot/files/5866246/Shadow.zip)

Accurate Behavior Prior to This PR:

![Vulkan Shadow Before](https://user-images.githubusercontent.com/50302479/105700864-c8044700-5f09-11eb-8604-2c99c4e402d9.PNG)

akien-mga commented:
Thanks!

JFonS commented:
@mrjustaguy  Looks like the shadow masks is not being taken into account, so the origin lines are casting a shadow.

reduz commented:
Indeed it looks like masking code or the cast shadow property is not being respected.. but I did not touch anything related to that in this PR so my guess is that it happened in an earlier one.

Xartorx commented:
@JFonS @reduz it's probably [this](https://github.com/godotengine/godot/issues/45109) bug, or somewhat related bug.

mrjustaguy commented:
You're Right, however for some reason it only manifests in this PR for me (Haven't seen it on any others thus far).. I removed the sphere and it was still shadowing the Line.. Refreshing the shadows however seems to Fix it
